include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -1056,14 +1056,22 @@ irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr
 			// ehh this whole thing of managing COATING is bad af
 			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
 			vec3 eta = bsdf_data_decodeIoR(bsdf_data, OP_COATING)[0];
-			float dummy;
 			vec3 fresnel = irr_glsl_fresnel_dielectric_frontface_only(eta, currInteraction.isotropic.NdotV);
 			float w = dot(fresnel, CIE_XYZ_Luma_Y_coeffs);
-			bool choseCoat = irr_glsl_partitionRandVariable(w, u.z, dummy);
+			float rcpProb;
+			bool choseCoat = irr_glsl_partitionRandVariable(w, u.z, rcpProb);
+			//TODO those weight multiplications are probably wrong, not needed or should concern remainder only
 			if (choseCoat)
+			{
+				weight *= (1.0-w);
 				break;
+			}
 			else
+			{
+				weight *= w;
+				//weight *= diffuse fresnel correction
 				ix = ix+1u;
+			}
```

Comment:
proper way to do it (putting asside the do we put weight on PDF issue)
```glsl
weight /= rcpProb;
if (choseCoat)
    break;
ix++;
```

ALWAYS use the `rcpProb` from the partition function!

Reply:
this is an awful if statement anyway... will be deleted, resolved